site/documentation/content/user-guide/coap-adapter.md

```diff
@@ -0,0 +1,230 @@
+
+The examples below refer to devices `4711` and `gw-1` of tenant `DEFAULT_TENANT` using *auth-ids* `sensor1` and `gw1` and corresponding passwords. The example deployment as described in the [Deployment Guides]({{< relref "deployment" >}}) comes pre-configured with the corresponding entities in its device registry component.
+Please refer to the [Credentials API]({{< relref "/api/credentials#standard-credential-types" >}}) for details regarding the different types of secrets.
+
+**NB** There is a subtle difference between the *device identifier* (*device-id*) and the *auth-id* a device uses for authentication. See [Device Identity]({{< relref "/concepts/device-identity.md" >}}) for a discussion of the concepts.
+
+## Message Limits
+
+Before accepting any telemetry or event or command messages, the CoAP adapter verifies that the configured [message limit]({{< relref "/concepts/resource-limits#messages-limit" >}}) is not exceeded.
+For CoAP currently only the messages limit is verified, the connections limit is not available. If the limit is exceeded then the incoming message is discarded with the status code `429 Too Many Requests`. 
+
+## CoAP Content Formats
+
+CoAP doesn't use a textual identifier for content types. Instead numbers are used, which are maintained by the [IANA](https://www.iana.org/).
+For the currently assigned numbers, please check [IANA - CoAP Content Formats](https://www.iana.org/assignments/core-parameters/core-parameters.xhtml#content-formats).
+Content types that do not map to one of the registered CoAP content formats are ignored.
+
+## Publish Telemetry Data (authenticated Device)
+
+The device is authenticated using PSK.
+
+* URI: `/telemetry`
+* Method: `POST`
+* Type:
+  * `CON`: *at least once* (`1`) QoS levels
+  * `NON`: *at most once* (`0`) QoS levels
+* Request Options:
+  * (optional) `content-format`: The type of payload contained in the request body. Required, if request contains payload.
+  * (optional) `URI-query: hono-ttd`: The number of seconds the device will wait for the response.
+  * (optional) `URI-query: empty`: If the request doesn't contain payload.
+* Request Body:
+  * (optional) Arbitrary payload encoded according to the given content type. Maybe empty, if `URI-query: empty` is provided.
+* Response Options:
+  * (optional) `content-format`: A media type describing the semantics and format of payload contained in the response body. This option will only be present if the response contains a command to be executed by the device which requires input data. If the command's content type do not map to one of the registered CoAP content formats, the option will not be present.
+  * (optional) `location-query: hono-command`: The name of the command to execute. This option will only be present if the response contains a command to be executed by the device.
```

Comment:
I have to admit that I am not really fond of abusing existing option codes for conveying information that is not at all related to the option's original semantics. This brings me back to introducing our own content format and option codes in the > 65000 range. For the options, this merely requires documenting them here. For the content-formats, we could create a wrapper around Cf's `MediaTypeRegistry` class which contains these codes as well ...
WDYT?

Reply:
Do you mean a "custom option" instead of the location?
Or a "custom content format", which structures the response's payload in "response-URI" and "command-payload"?

Reply:
Both, using a custom option instead of `location-query` and `location-path` as well as using custom content-format codes for e.g. `application/vnd.eclipse-hono-empty-notification`.

Reply:
Both requires, that a implementation is able to process such "custom" values. I just don't know, if that is the case.

Reply:
When you say *implementation*, what are you referring to? A device? A backend application?

Reply:
A embedded device. At least, this should be checked for "libcoap" and "erbium". And for the content-format, californium must be adapted.
I would prefer to raise a question in the ietf core-mailing list about the usage of the "location". For me this case looks pretty close to the original intention of the "location".

Reply:
> And for the content-format, californium must be adapted.And for the content-format, californium must be adapted.

IMHO we can simply wrap the `MediaTypeRegistry` and use our wrapper instead in the CoAP adapter code. Don't you think?

Reply:
> At least, this should be checked for "libcoap" and "erbium"

Can you do that?

Reply:
I have only 1 day left before my vacation and I'm sorry, but this will not be possible before that.

I want to release Californium 2.2.0-RC1 tomorrow, and so I'm currently just too busy.

Reply:
We don't need to do that as part of this PR. This is just about documenting the current status quo, i.e. the way it is implemented at the moment. However, I wanted to start a discussion around whether it should stay that way.

Reply:
OK, but as I wrote, I would prefer to ask the ietf core list about their opinion using the location option. I hope, I can write to the list tomorrow.

Reply:
Reading [2.01 Created](https://tools.ietf.org/html/rfc7252#section-5.9.1.1) my impression is, that the "location" is used to control a cache. Assuming no cache, it doesn't harm.